File: frontend/types/research.ts
/**
 * Research-related TypeScript types.
 */

export type ResearchStatus =
  | "created"
  | "clarifying"
  | "planning"
  | "ready"
  | "executing"
  | "synthesizing"
  | "complete"
  | "error";

export interface ResearchTask {
  source: string;
  description: string;
  query: string;
  expected_output: string;
  status: string;
}

export interface ResearchPhase {
  name: string;
  description: string;
  estimated_time_minutes: number;
  tasks: ResearchTask[];
  status: string;
}

export interface ResearchPlan {
  title: string;
  objective: string;
  estimated_time_minutes: number;
  phases: ResearchPhase[];
  deliverables: string[];
}

export interface Message {
  role: "user" | "assistant";
  content: string;
  timestamp: string;
}

export interface ResearchSession {
  id: string;
  query: string;
  status: ResearchStatus;
  created_at: string;
  updated_at?: string;
  messages: Message[];
  plan?: ResearchPlan;
  has_results: boolean;
  has_report: boolean;
  error?: string;
}

export interface Visualization {
  type: "bar_chart" | "line_chart" | "pie_chart" | "table" | "metric" | "area_chart";
  title: string;
  description: string;
  data: Record<string, unknown>;
}

export interface ReportSection {
  title: string;
  content: string;
  visualizations: Visualization[];
  citations: string[];
}

export interface Report {
  id: string;
  title: string;
  executive_summary: string;
  sections: ReportSection[];
  key_findings: string[];
  recommendations: string[];
  generated_at: string;
  metadata: Record<string, unknown>;
}

// SSE Event Types
export interface ChatDeltaEvent {
  type: "chat_delta";
  content: string;
}

export interface StatusChangeEvent {
  type: "status_change";
  status: ResearchStatus;
}

export interface PlanningStartEvent {
  type: "planning_start";
  message: string;
}

export interface PlanReadyEvent {
  type: "plan_ready";
  plan: ResearchPlan;
}

export interface ExecutionStartEvent {
  type: "execution_start";
  total_phases: number;
}

export interface PhaseStartEvent {
  type: "phase_start";
  phase_index: number;
  phase_name: string;
  tasks: string[];
}

export interface TaskCompleteEvent {
  type: "task_complete";
  phase_index: number;
  task: string;
  success: boolean;
  result_count: number;
}

export interface PhaseCompleteEvent {
  type: "phase_complete";
  phase_index: number;
}

export interface SynthesisStartEvent {
  type: "synthesis_start";
}

export interface ReportReadyEvent {
  type: "report_ready";
  report: Report;
}

export interface ErrorEvent {
  type: "error";
  message: string;
}

export type SSEEvent =
  | ChatDeltaEvent
  | StatusChangeEvent
  | PlanningStartEvent
  | PlanReadyEvent
  | ExecutionStartEvent
  | PhaseStartEvent
  | TaskCompleteEvent
  | PhaseCompleteEvent
  | SynthesisStartEvent
  | ReportReadyEvent
  | ErrorEvent;

File: backend/app/models/research.py
"""Research-related data models."""

from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class ResearchStatus(str, Enum):
    """Status of a research session."""

    CREATED = "created"
    CLARIFYING = "clarifying"
    PLANNING = "planning"
    READY = "ready"
    EXECUTING = "executing"
    SYNTHESIZING = "synthesizing"
    COMPLETE = "complete"
    ERROR = "error"


class ResearchTask(BaseModel):
    """A single research task within a phase."""

    source: str = Field(..., description="Data source to query")
    description: str = Field(..., description="What this task does")
    query: str = Field(..., description="Search query or parameters")
    expected_output: str = Field(default="", description="Expected data type")
    status: str = Field(default="pending", description="Task status")

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary."""
        return self.model_dump()


class ResearchPhase(BaseModel):
    """A phase of the research plan containing multiple tasks."""

    name: str = Field(..., description="Phase name")
    description: str = Field(default="", description="Phase description")
    estimated_time_minutes: int = Field(default=2, description="Estimated duration")
    tasks: List[ResearchTask] = Field(default_factory=list)
    status: str = Field(default="pending", description="Phase status")

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary."""
        return {
            "name": self.name,
            "description": self.description,
            "estimated_time_minutes": self.estimated_time_minutes,
            "tasks": [t.to_dict() for t in self.tasks],
            "status": self.status,
        }


class ResearchPlan(BaseModel):
    """A complete research plan."""

    title: str = Field(..., description="Plan title")
    objective: str = Field(default="", description="Research objective")
    estimated_time_minutes: int = Field(default=10, description="Total estimated time")
    phases: List[ResearchPhase] = Field(default_factory=list)
    deliverables: List[str] = Field(default_factory=list)

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary."""
        return {
            "title": self.title,
            "objective": self.objective,
            "estimated_time_minutes": self.estimated_time_minutes,
            "phases": [p.to_dict() for p in self.phases],
            "deliverables": self.deliverables,
        }


class Message(BaseModel):
    """A chat message."""

    role: str = Field(..., description="Message role: user or assistant")
    content: str = Field(..., description="Message content")
    timestamp: datetime = Field(default_factory=datetime.utcnow)

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary."""
        return {
            "role": self.role,
            "content": self.content,
            "timestamp": self.timestamp.isoformat(),
        }


class ResearchSession(BaseModel):
    """A complete research session."""

    id: str = Field(..., description="Session ID")
    query: str = Field(..., description="Original research query")
    status: ResearchStatus = Field(default=ResearchStatus.CREATED)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: Optional[datetime] = None
    messages: List[Message] = Field(default_factory=list)
    plan: Optional[ResearchPlan] = None
    results: Optional[Dict[str, List[Dict[str, Any]]]] = None
    report: Optional[Any] = None  # Will be Report type
    error: Optional[str] = None

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary."""
        return {
            "id": self.id,
            "query": self.query,
            "status": self.status.value,
            "created_at": self.created_at.isoformat(),
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
            "messages": [m.to_dict() for m in self.messages],
            "plan": self.plan.to_dict() if self.plan else None,
            "has_results": self.results is not None,
            "has_report": self.report is not None,
            "error": self.error,
        }


class TaskResult(BaseModel):
    """Result from executing a single research task."""

    task_name: str = Field(..., description="Name of the task")
    source: str = Field(..., description="Data source used")
    success: bool = Field(..., description="Whether task succeeded")
    data: Optional[List[Dict[str, Any]]] = Field(default=None)
    error: Optional[str] = Field(default=None)
    execution_time_ms: int = Field(default=0)

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary."""
        return {
            "task_name": self.task_name,
            "source": self.source,
            "success": self.success,
            "result_count": len(self.data) if self.data else 0,
            "error": self.error,
            "execution_time_ms": self.execution_time_ms,
        }


class ExecutionLog(BaseModel):
    """Log entry for research execution."""

    timestamp: datetime = Field(default_factory=datetime.utcnow)
    phase: str = Field(..., description="Current phase")
    task: Optional[str] = Field(default=None)
    status: str = Field(..., description="Status: started, completed, error")
    message: str = Field(default="")
    data: Optional[Dict[str, Any]] = None

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary."""
        return {
            "timestamp": self.timestamp.isoformat(),
            "phase": self.phase,
            "task": self.task,
            "status": self.status,
            "message": self.message,
            "data": self.data,
        }
